### PR TITLE
fix(shared): improve types of cx

### PR DESCRIPTION
**Pull request by Haroenv**
The actual type of cx is more complex than `string | undefined`, as it's only undefined if *all* items are falsy. This is quite tough to express in TypeScript, hence the fairly complex change + casting. I couldn't simplify it further.

This PR also had to update a little bit of tooling (I tried to do the minimal), because the eslint parser choked on the spread of a tuple in an extend

closes #21

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -36,6 +37,7 @@ module.exports = {
         allow: ['hierarchical_categories'],
       },
     ],
+    '@typescript-eslint/no-use-before-define': OFF,
```
Comment (Haroenv) [inline .eslintrc.js:40]: it's buggy with this version of the plugin (and we don't really need it here)

File: packages/horizontal-slider-vdom/src/types/HorizontalSlider.ts [hunk 2]
```diff
@@ -22,6 +22,6 @@ export type HorizontalSliderTranslations = Partial<{
 export type HorizontalSliderProps<TItem extends RecordWithObjectID> = {
   classNames?: HorizontalSliderClassnames;
   items: TItem[];
-  itemComponent({ item: TItem }): JSX.Element;
+  itemComponent(_props: { item: TItem }): JSX.Element;
```
Comment (Haroenv) [inline packages/horizontal-slider-vdom/src/types/HorizontalSlider.ts:25]: yay, unrelated bug fix thanks to the update of eslint

File: packages/shared/src/cx.ts [hunk 3]
```diff
@@ -1,15 +1,55 @@
...
+        : Join<TTail, TSeparator, `${TResult}${TSeparator}${THead}`>
+      : string
+    : string
+  : string;
+
+type FallbackToUndefined<TValue> = TValue extends '' ? undefined : TValue;
+
+type CX<
+  TClasses extends Array<ClassValue | ClassValue[]>
+> = FallbackToUndefined<Join<RemoveFalsy<Flat<Flat<TClasses>>>, ' '>>;
```
Comment (Haroenv) [inline packages/shared/src/cx.ts:41]: the actual important part of this PR

Comment (Haroenv) [reply packages/shared/src/cx.ts:41]: in theory, if we make a (breaking) change that outputs `string` always (instead of falling back to undefined, avoiding an empty class being written), none of these things are needed, just not sure if it counts as a breaking change to no longer return undefined, it was clearly on purpose in the initial implementation

**Comment by francoischalifour** [comment]
I'm wondering if it would have been simpler to accommodate InstantSearch to UI components. Joining an array with spaces is OK to keep the `cx()` API simple. We're leaking the InstantSearch API here.

**Comment by Haroenv** [comment]
I tried that, but it seemed odd to need to deal with classnames being optional in 200+ places in InstantSearch, the one which makes more sense to me is to remove the `|| undefined` branch
